.github/workflows/snek.yml

```diff
@@ -0,0 +1,21 @@
+name: Snek
+
+on: [push]
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+    
+    steps:
+    - uses: actions/checkout@v1
+    - name: install dependencies
+      run: sudo apt install gcc-avr avr-libc python3-serial gcc-arm-none-eabi libreadline-dev
+    - name: install lola 
+      run: |
+        git clone https://github.com/keith-packard/lola.git
+        cd lola && make && sudo make install
+    - name: posix
+      run: |
+       make -C ports/posix
```

Comment:
If you're just running the posix version, you don't need any of the embedded toolchains.

Reply:
built posix only for testing, the idea is to build them all.

Reply:
Hrm. That's a great goal, but will require that we get current toolchains. Can we install .deb packages on the target machine from another repo? The C library should (eventually) make it out of the Debian new queue, at which point everything will be in Debian unstable. Alternatively, we can pull things from my home repository too.

Reply:
Figured out how to use a custom OS with packages we need starting from debian unstable.

Installed picolibc from your archive on a debian unstable, but I am getting errors when build ARM and RISC-V based ports:

from feather port:

```
root@398da460ce06:/snek/ports/feather# make
  CC  snek-feather-1.2.elf
/usr/lib/gcc/arm-none-eabi/7.3.1/../../../arm-none-eabi/bin/ld: cannot find /usr/lib/picolibc/arm-none-eabi/lib/thumb/v6-m/crt0.o: No such file or directory
collect2: error: ld returned 1 exit status
make: *** [Makefile:44: snek-feather-1.2.elf] Error 1
```
similar for qemu-arm port:

```
root@398da460ce06:/snek/ports/qemu-arm# make
  CC  snek-qemu-arm-1.2.elf
/usr/lib/gcc/arm-none-eabi/7.3.1/../../../arm-none-eabi/bin/ld: cannot find /usr/lib/picolibc/arm-none-eabi/lib/thumb/v7-m/crt0.o: No such file or directory
collect2: error: ld returned 1 exit status
make: *** [Makefile:33: snek-qemu-arm-1.2.elf] Error 1
```

from qemu-risc port:

```
root@398da460ce06:/snek/ports/qemu-riscv# make
  CC  snek-qemu-riscv-1.2.elf
/usr/lib/riscv64-unknown-elf/bin/ld: cannot find /usr/lib/picolibc/arm-none-eabi/lib/rv32imac/ilp32/crt0.o: No such file or directory
/usr/lib/riscv64-unknown-elf/bin/ld: skipping incompatible /usr/lib/picolibc/arm-none-eabi/lib/libm.a when searching for -lm
/usr/lib/riscv64-unknown-elf/bin/ld: cannot find -lm
/usr/lib/riscv64-unknown-elf/bin/ld: skipping incompatible /usr/lib/picolibc/arm-none-eabi/lib/libc.a when searching for -lc
/usr/lib/riscv64-unknown-elf/bin/ld: cannot find -lc
/usr/lib/riscv64-unknown-elf/bin/ld: skipping incompatible /usr/lib/picolibc/arm-none-eabi/lib/libsemihost.a when searching for -lsemihost
/usr/lib/riscv64-unknown-elf/bin/ld: cannot find -lsemihost
collect2: error: ld returned 1 exit status
make: *** [Makefile:33: snek-qemu-riscv-1.2.elf] Error 1
```

from `dpkg -L picolibc-arm-none-eabi | grep crt0`:

```
/usr/lib/picolibc/arm-none-eabi/lib/arm/v5te/hard/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/arm/v5te/softfp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/nofp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v6-m/nofp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v7/nofp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v7+fp/hard/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v7+fp/softfp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v7-m/nofp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v7e-m/nofp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v7e-m+dp/hard/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v7e-m+dp/softfp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v7e-m+fp/hard/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v7e-m+fp/softfp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v8-m.base/nofp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v8-m.main/nofp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v8-m.main+dp/hard/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v8-m.main+dp/softfp/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v8-m.main+fp/hard/crt0.o
/usr/lib/picolibc/arm-none-eabi/lib/thumb/v8-m.main+fp/softfp/crt0.o
```

and for risc-v:

from `dpkg -L picolibc-riscv64-unknown-elf | grep crt0`:

```
/usr/lib/picolibc/riscv64-unknown-elf/lib/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv32e/ilp32e/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv32eac/ilp32e/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv32em/ilp32e/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv32emac/ilp32e/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv32i/ilp32/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv32iac/ilp32/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv32im/ilp32/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv32imac/ilp32/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv32imafc/ilp32f/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv32imafdc/ilp32d/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv32imf/ilp32f/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv64i/lp64/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv64iac/lp64/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv64im/lp64/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv64imac/lp64/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv64imafc/lp64f/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv64imafdc/lp64d/crt0.o
/usr/lib/picolibc/riscv64-unknown-elf/lib/rv64imf/lp64f/crt0.o
```
I can open separate issues if we need, after figure this out I think I can build all ports on CI.

Reply:
Also had to adjust specs file path as gcc was unable to locate it as is.

```diff
index ec3d80f..9834307 100644
--- a/chips/qemu/snek-qemu.defs
+++ b/chips/qemu/snek-qemu.defs
@@ -31,7 +31,7 @@ SNEK_LOCAL_VPATH = $(SNEK_QEMU)
 PICOLIBC_PRINTF_CFLAGS =

 PICOLIBC_CFLAGS= \
-       -specs=picolibc.specs \
+       -specs=/usr/lib/gcc/arm-none-eabi/8.3.1/picolibc.specs \
        $(PICOLIBC_PRINTF_CFLAGS)

 OPT=-O0 -g
diff --git a/chips/samd21/snek-samd21.defs b/chips/samd21/snek-samd21.defs
index 2889e83..0c50e07 100644
--- a/chips/samd21/snek-samd21.defs
+++ b/chips/samd21/snek-samd21.defs
@@ -66,7 +66,7 @@ SNEK_SAMD21_BUILTINS = \
 PICOLIBC_PRINTF_CFLAGS =

 PICOLIBC_CFLAGS= \
-       -specs=picolibc.specs \
+       -specs=/usr/lib/gcc/arm-none-eabi/8.3.1/picolibc.specs \
        $(PICOLIBC_PRINTF_CFLAGS)

 OPT=-Os
```

Reply:
Are you using debian unstable versions of the gcc ports? It looks like the path being computed by the linker

`/usr/lib/picolibc/arm-none-eabi/lib/thumb/v6-m/crt0.o` 

is not the same as the package installs 

`/usr/lib/picolibc/arm-none-eabi/lib/thumb/v6-m/nofp/crt0.o`

picolibc.specs uses '%M' to add the architecture-specific portion of the library path so that the linker finds the right libraries, but if %M comes from a differently configured compiler than picolibc was built with, then it won't find the files.

If you're already using gcc from debian unstable, then I'm not sure what's going on.

Reply:
Oh. If the compiler isn't the *exact same version* then the .specs file will get installed in the wrong place.

I'll go work on picolibc packaging some more. I discovered this problem when trying to cross-compile snek for i386 where the arm gcc version is 7.3.1 (!)

Reply:
oh! I'm using a *newer version of gcc* on my machine because I'm crazy. I wonder why this hasn't been uploaded to unstable yet. Sorry for the confusion; I'll see what I can do :-)

Reply:
gcc used on tests above from debian unstable on docker `gcc (Debian 9.2.1-19) 9.2.1 20191109`. We'll probably need to use a docker tag to get the specific version you use when you push the .deb package to your archive, not a big deal, just need to know the release day and get a tag from docker hub.

Reply:
Ok, I've redesigned the picolibc packaging so that it automagically installs picolibc.specs in the directory associated with the current gcc-arm-none-eabi and gcc-riscv64-unknown-elf packages. However, it won't adapt the path names of the various library bits, like crt0.o. It turns out I've got a custom version of gcc-arm-none-eabi in my environment as we've been having trouble getting the debian package updated and I was testing stuff a couple of months ago. I'll reset my world to use the current debian unstable compiler, which you can actually get at least. Building will take a while...

Reply:
Phew. I *think* the updated picolibc packages in my home repository are finally working now. I'm building with arm-none-eabi-gcc version 7.3.1 20180622 and have snek successfully compiling. That should be the version you are getting, if not, it's in debian unstable.

Next up is to add file i/o support to snek-qemu-arm so we can run that under qemu and test snek on another architecture. That should be fun :-)

Thanks for your patience; I've been waiting for the Debian gcc-arm-none-eabi package to be updated for several months now and had been testing the new version locally (hence all of this adventure).